File: apps/api/src/common/screen-context.test.ts
import { describe, it, expect } from "vitest";
import { WORKER_FEEDBACK_SCREEN_MAX } from "@badabhai/types";

import { sanitizeScreenContext } from "./screen-context";

/**
 * `sanitizeScreenContext` is the ONLY thing standing between an untrusted client and three
 * sinks that must never hold an identifier: the `worker_feedback` row, the `feedback.submitted`
 * event, and the API log line. So this file is written around two questions, in this order.
 *
 *  1. DOES IT STRIP IDENTIFIERS? A path is the natural carrier for one — every entity on this
 *     platform is a uuid in a URL — and the whole reason the column stores a PATTERN is that
 *     `/jobs/<uuid>` links a feedback row to one specific job. A normalizer that missed a shape
 *     would put that link on the event spine, where §2 forbids it, and nothing downstream would
 *     notice because the value would still look like a route.
 *
 *  2. DOES IT REFUSE RATHER THAN THROW? Every rejection here must be a `null`, because the
 *     caller is on the request path of a worker's typed feedback and a throw would cost them a
 *     paragraph over a field they never filled in.
 */

describe("sanitizeScreenContext — identifiers never survive", () => {
  it("replaces a uuid segment with :id", () => {
    expect(sanitizeScreenContext("/jobs/6f2c04e0-4f89-41d3-9a0c-0305e82c3301/apply")).toBe(
      "/jobs/:id/apply",
    );
  });

  it("replaces EVERY id in a path, not only the first", () => {
    // A single-substitution implementation (a non-global replace over the whole string) passes
    // the test above and leaks the second id — which on this platform is the session id.
    expect(
      sanitizeScreenContext(
        "/workers/6f2c04e0-4f89-41d3-9a0c-0305e82c3301/sessions/11111111-2222-4333-8444-555555555555",
      ),
    ).toBe("/workers/:id/sessions/:id");
  });

  it("replaces an all-numeric segment", () => {
    // Not every id on a screen is a uuid: page indices and legacy numeric ids are both real,
    // and `/orders/91723` identifies exactly one thing.
    expect(sanitizeScreenContext("/orders/91723")).toBe("/orders/:id");
  });

  it("substitutes a uuid whose version nibble is not one we mint today", () => {
    // The regex deliberately does not pin `[89ab]`/`4`: a v7 uuid, or one from a client that
    // generated it loosely, is still an IDENTIFIER. Recognising only today's shapes is how a
    // normalizer quietly stops working.
    expect(sanitizeScreenContext("/jobs/018f4c7a-9b21-7c3d-0e5f-0a1b2c3d4e5f")).toBe("/jobs/:id");
  });

  it("drops the query string — the likeliest place a client parks worker input", () => {
    expect(sanitizeScreenContext("/search?q=welder%20mumbai&page=2")).toBe("/search");
  });

  it("drops the fragment as well as the query, in whichever order they appear", () => {
    // ⚠ NOT AN ORDER ASSERTION, and it used to claim to be one. Each cut keeps the prefix
    // before its own delimiter, so `#`-then-`?` and `?`-then-`#` produce the same string for
    // every input — swapping the two lines in the implementation reddens nothing, measured.
    // What IS worth pinning is that neither delimiter survives whichever comes first.
    expect(sanitizeScreenContext("/profile#section?tab=skills")).toBe("/profile");
    expect(sanitizeScreenContext("/profile?tab=skills#section")).toBe("/profile");
  });

  /**
   * ⚠ THE BUG THIS SUITE MISSED, AND THE REASON IT MISSED IT.
   *
   * Every case above puts the identifier alone in its segment. The first implementation tested
   * `^<uuid>$` / `^\d+$` PER SEGMENT, so all of them passed while an id sharing its segment
   * with one other character went through verbatim — into the row, onto `feedback.submitted`,
   * and into the API log line. Each input below was executed against the shipped code and came
   * back unchanged; each is a §2 violation on three sinks at once.
   */
  describe("an identifier does not have to be the WHOLE segment", () => {
    it.each([
      ["a uuid behind a prefix", "/jobs/id-6f2c04e0-4f89-41d3-9a0c-0305e82c3301/apply"],
      ["a uuid with a trailing character", "/jobs/6f2c04e0-4f89-41d3-9a0c-0305e82c3301x/apply"],
      ["the dash-less uuid form", "/jobs/6f2c04e04f8941d39a0c0305e82c3301/apply"],
      ["a phone number and a name", "/w/9876543210-ravi"],
      ["a name and a phone number", "/support/ravi-kumar-9876543210"],
      ["a grouped 12-digit number", "/AADHAAR/1234-5678-9012"],
      ["a phone after a scheme-ish colon", "/tel:919876543210"],
      ["dot-separated worker detail", "/ramesh.kumar.9876543210"],
      ["an alphanumeric id", "/chat/AbCdEf123456/msg"],
    ])("substitutes %s", (_label, raw) => {
      const out = sanitizeScreenContext(raw);
      expect(out, raw).not.toBeNull();
      // No uuid, no dash-less hex id, and no digit run long enough to be a number about a
      // person. The assertion is on the OUTPUT so it keeps holding for shapes nobody has
      // written a case for.
      expect(out!, raw).not.toMatch(
        /[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}/i,
      );
      expect(out!, raw).not.toMatch(/[0-9]{4,}/);
      expect(out!, raw).not.toMatch(/(?=[0-9a-fA-F]*[0-9])[0-9a-fA-F]{16,}/);
    });

    it("substitutes EVERY run in one segment, not only the first", () => {
      expect(sanitizeScreenContext("/x/9876543210-and-1234567890")).toBe("/x/:id-and-:id");
    });

    /**
     * ⚠ THE RESIDUAL, ASSERTED SO IT IS NOT MISTAKEN FOR A GUARANTEE. An opaque token that is
     * neither hex nor digits is indistinguishable from a route word by any structural rule, so
     * it survives — and no comment on this field may claim otherwise. Closing this needs an
     * ALLOWLIST of the client's finite route table, not a longer denylist.
     */
    it("does NOT catch an opaque non-hex token — a denylist cannot", () => {
      expect(sanitizeScreenContext("/u/dGVzdEBleGFtcGxlLmNvbQ")).toBe(
        "/u/dGVzdEBleGFtcGxlLmNvbQ",
      );
    });

    /** A long lowercase word made only of `a`–`f` is hex by charset and is not an id. */
    it("does not mistake a long hex-lettered WORD for an id", () => {
      expect(sanitizeScreenContext("/deadbeefdeadbeef/edit")).toBe("/deadbeefdeadbeef/edit");
    });
  });

  it("never returns a value containing a uuid or a bare digit run", () => {
    // The structural version of the assertions above, so it keeps holding for an input shape
    // nobody has written a case for yet.
    const inputs = [
      "/jobs/6f2c04e0-4f89-41d3-9a0c-0305e82c3301",
      "/a/1/b/22/c/333",
      "/chat/6f2c04e0-4f89-41d3-9a0c-0305e82c3301?from=/jobs/9",
      "/workers/6F2C04E0-4F89-41D3-9A0C-0305E82C3301/profile",
    ];
    for (const input of inputs) {
      const out = sanitizeScreenContext(input);
      expect(out, input).not.toBeNull();
      expect(out!, input).not.toMatch(/[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}/i);
      expect(out!.split("/"), input).not.toContainEqual(expect.stringMatching(/^\d+$/));
    }
  });
});

describe("sanitizeScreenContext — sanitize, never reject, never throw", () => {
  it.each([
    ["a non-string", 42],
    ["null", null],
    ["undefined", undefined],
    ["an object", { path: "/jobs" }],
    ["an array (a repeated field)", ["/jobs", "/home"]],
    ["an empty string", ""],
    ["whitespace only", "   "],
    ["an unrooted path", "jobs/apply"],
    ["a full URL", "https://badabhai.ai/jobs"],
    // A rooted path by every naive check, and a SCHEME-RELATIVE URL to a host we do not
    // control — rendered as a link on the admin screen it is an outbound navigation an
    // attacker chose. The `//` arm of the shared pattern is what refuses it.
    ["a scheme-relative URL", "//evil.example/jobs"],
    ["a doubled slash anywhere", "/jobs//apply"],
    ["markup", "/jobs/<script>alert(1)</script>"],
    ["whitespace inside", "/jobs/my job"],
    ["percent-encoding", "/jobs/%2e%2e%2fadmin"],
    ["a newline", "/jobs\n/admin"],
    ["non-ASCII", "/नौकरी/आवेदन"],
  ])("returns null for %s", (_label, raw) => {
    expect(sanitizeScreenContext(raw)).toBeNull();
  });

  /**
   * The pre-split DoS bound is LOSSY, and the comment that used to sit on it claimed it was
   * not ("cannot discard anything the check below would have kept"). Pinned here so the false
   * invariant cannot come back as a justification for raising the multiplier.
   */
  it("discards a value past the raw DoS bound even though it would normalize small", () => {
    const raw = `/orders/${"9".repeat(1300)}`;
    expect(raw.length).toBeGreaterThan(WORKER_FEEDBACK_SCREEN_MAX * 10);
    // Its normalized form would be `/orders/:id` — eleven characters, well inside the bound.
    expect(sanitizeScreenContext(raw)).toBeNull();
  });

  it("returns null past the bound rather than truncating", () => {
    // Truncating would produce a route pattern that names a screen nobody can navigate to, and
    // it would look authoritative on the admin list. Null reads as "unknown screen", which is
    // the honest answer.
    expect(sanitizeScreenContext(`/${"a".repeat(WORKER_FEEDBACK_SCREEN_MAX)}`)).toBeNull();
    expect(sanitizeScreenContext(`/${"a".repeat(WORKER_FEEDBACK_SCREEN_MAX - 1)}`)).not.toBeNull();
  });

  it("measures the bound AFTER substitution, so an id-heavy path is not lost to its ids", () => {
    // Four uuid segments is 148 raw characters — over the bound — and 20 once normalized. The
    // deep screens are exactly the ones a "button kaam nahi kar raha" report is about, so
    // measuring before substitution would discard the reports that matter most.
    const raw = "/a/6f2c04e0-4f89-41d3-9a0c-0305e82c3301".repeat(4);
    expect(raw.length).toBeGreaterThan(WORKER_FEEDBACK_SCREEN_MAX);
    expect(sanitizeScreenContext(raw)).toBe("/a/:id/a/:id/a/:id/a/:id");
  });

  it("NEVER throws, for any of these", () => {
    for (const raw of [undefined, null, 0, NaN, {}, [], "/", "/".repeat(5000), "?", "#"]) {
      expect(() => sanitizeScreenContext(raw)).not.toThrow();
    }
  });

  it("keeps a plain route unchanged, including its root and a trailing slash", () => {
    expect(sanitizeScreenContext("/")).toBe("/");
    expect(sanitizeScreenContext("/home")).toBe("/home");
    expect(sanitizeScreenContext("/jobs/")).toBe("/jobs/");
    expect(sanitizeScreenContext("  /settings/notifications  ")).toBe("/settings/notifications");
  });
});

File: apps/worker-app/test/features/feedback/screen_context_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:badabhai_worker_app/features/feedback/domain/screen_context.dart';

/// THE CLIENT HALF OF A CROSS-LANGUAGE CONTRACT.
///
/// `normalizeScreenContext` (Dart) and `sanitizeScreenContext` (TypeScript,
/// `apps/api/src/common/screen-context.ts`) must make the SAME decision about the
/// same string. Nothing in either build can check that — there is no shared test
/// runner and no shared type — so this file pins the same constant and, case for
/// case, the same EXAMPLE STRINGS as `apps/api/src/common/screen-context.test.ts`.
/// A repo that has been bitten by a mock which accepted anything does not get to
/// call "the server normalizes too" a test.
///
/// And it is written for what the normalizer PERMITS as much as for what it
/// strips: a rule that rejects everything passes every "does it catch X" test and
/// is discovered only when it silently deletes the screen label off every real
/// report.
void main() {
  test('the bound is the SERVER\'s number, pinned as a literal', () {
    // WORKER_FEEDBACK_SCREEN_MAX in packages/types, and the
    // `worker_feedback_screen_context_len_chk` CHECK. Every other test in this
    // file reads the constant, so none of them can catch the constant itself
    // being wrong — which is exactly how a hand-copied cross-language number
    // drifts.
    expect(kWorkerFeedbackScreenMax, 128);
  });

  group('the four shapes the contract turns on', () {
    // ── 1. A UUID SEGMENT. Every entity id on this platform is a uuid in a
    // path, and `/jobs/<uuid>/apply` ties a feedback row to ONE specific job —
    // a fact about the worker nobody asked them to disclose, and the reason the
    // stored value is a pattern at all.
    test('a uuid segment becomes :id', () {
      expect(
        normalizeScreenContext(
            '/jobs/6f2c04e0-4f89-41d3-9a0c-0305e82c3301/apply'),
        '/jobs/:id/apply',
      );
    });

    // ── 2. AN ALL-NUMERIC SEGMENT. Not every id is a uuid.
    test('an all-numeric segment becomes :id', () {
      expect(normalizeScreenContext('/orders/91723'), '/orders/:id');
    });

    // ── 3. A QUERY STRING — the likeliest place a client parks worker input.
    test('the query string is dropped', () {
      expect(
        normalizeScreenContext('/search?q=welder%20mumbai&page=2'),
        '/search',
      );
    });

    // ── 4. AN ORDINARY ROUTE, WHICH MUST SURVIVE UNCHANGED. This is the case
    // that stops the whole feature from quietly becoming "screen: unknown" on
    // every row — a normalizer that only ever strips is indistinguishable from a
    // normalizer that always returns null until someone reads the admin list.
    test('an ordinary route survives byte for byte', () {
      expect(normalizeScreenContext('/settings/notifications'),
          '/settings/notifications');
    });
  });

  group('identifiers never survive (mirrors screen-context.test.ts)', () {
    test('EVERY id in a path, not only the first', () {
      // A single-substitution implementation passes the uuid case above and leaks
      // the second id — which on this platform is the session id.
      expect(
        normalizeScreenContext(
          '/workers/6f2c04e0-4f89-41d3-9a0c-0305e82c3301'
          '/sessions/11111111-2222-4333-8444-555555555555',
        ),
        '/workers/:id/sessions/:id',
      );
    });

    test('a uuid whose version nibble is not one we mint today', () {
      // v7, or one a client generated loosely, is still an IDENTIFIER.
      expect(normalizeScreenContext('/jobs/018f4c7a-9b21-7c3d-0e5f-0a1b2c3d4e5f'),
          '/jobs/:id');
    });

    test('an UPPERCASE uuid', () {
      expect(
        normalizeScreenContext(
            '/workers/6F2C04E0-4F89-41D3-9A0C-0305E82C3301/profile'),
        '/workers/:id/profile',
      );
    });

    test('BOTH the fragment and the query are dropped, in either order', () {
      // What is load-bearing is that BOTH cuts exist — not their order. Measured:
      // swapping them changes no result, because each cut keeps the prefix before
      // its marker, so whichever marker comes FIRST truncates the value either
      // way. The test that used to sit here asserted the order was load-bearing
      // and passed under both orders, which is a test that cannot fail.
      expect(normalizeScreenContext('/profile#section?tab=skills'), '/profile');
      expect(normalizeScreenContext('/a#b?c'), '/a');
      expect(normalizeScreenContext('/a?b#c'), '/a');
      expect(normalizeScreenContext('/a#b?c#d'), '/a');
      expect(normalizeScreenContext('/a?b#c?d'), '/a');
    });

    test('a long query string does not cost the route its name', () {
      // The pre-bound is measured against the CUT path, not the raw value. Held
      // against `raw` it discarded ordinary short routes: the query cut and the
      // trim are unbounded shrinks that run after it, so a 1281-character
      // `/jobs?<...>` normalized to null — "unknown screen" for a worker who was
      // plainly on /jobs.
      expect(normalizeScreenContext('/jobs?${'x' * 1275}'), '/jobs');
      expect(normalizeScreenContext('/jobs#${'x' * 1275}'), '/jobs');
      expect(normalizeScreenContext('/search?q=${'a' * 1400}'), '/search');
      expect(normalizeScreenContext('/jobs${' ' * 1300}'), '/jobs');
      // The bound still exists, on the part that survives the cuts.
      expect(normalizeScreenContext('/${'a' * 1300}'), isNull);
    });

    test('the charset is exactly what the server pattern admits', () {
      // A BACKSTOP for `_kScreenCharset`. Widening it is the drift that happens
      // the first time a route needs one of these, and it costs data on the
      // SERVER side (the value is stored NULL), which no client test would ever
      // see. Every one of these is a legal RFC 3986 path character that the
      // shared `WORKER_FEEDBACK_SCREEN_PATTERN` nonetheless refuses.
      const List<String> refused = <String>[
        '/jobs/~me', '/jobs/@me', '/jobs/a+b', '/jobs/a,b', '/jobs/a;b',
        r'/jobs/a$b', "/jobs/a'b", '/jobs/a(b', '/jobs/a)b', '/jobs/a*b',
        '/jobs/a=b', '/jobs/a!b', '/jobs/a&b', '/jobs/a[b', '/jobs/a]b',
        '/jobs/a^b', '/jobs/a`b', '/jobs/a{b', '/jobs/a}b', '/jobs/a|b',
        r'/jobs/a\b', '/jobs/a%20b', '/jobs/a b', '/jobs/a"b', '/jobs/a<b',
        '/jobs/a>b',
      ];
      for (final String input in refused) {
        expect(normalizeScreenContext(input), isNull, reason: input);
      }
      // ...and the ones it DOES admit stay admitted.
      expect(normalizeScreenContext('/a.b/C_d-e'), '/a.b/C_d-e');
      expect(normalizeScreenContext('/jobs/:id/apply'), '/jobs/:id/apply');
    });

    test('no output ever contains a uuid or a bare digit run', () {
      // The structural version, so it keeps holding for a shape nobody has
      // written a case for yet.
      const List<String> inputs = <String>[
        '/jobs/6f2c04e0-4f89-41d3-9a0c-0305e82c3301',
        '/a/1/b/22/c/333',
        '/chat/6f2c04e0-4f89-41d3-9a0c-0305e82c3301?from=/jobs/9',
        '/workers/6F2C04E0-4F89-41D3-9A0C-0305E82C3301/profile',
      ];
      final RegExp uuid = RegExp(
        r'[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}',
        caseSensitive: false,
      );
      for (final String input in inputs) {
        final String? out = normalizeScreenContext(input);
        expect(out, isNotNull, reason: input);
        expect(uuid.hasMatch(out!), isFalse, reason: input);
        expect(
          out.split('/').any((String s) => RegExp(r'^[0-9]+$').hasMatch(s)),
          isFalse,
          reason: input,
        );
      }
    });
  });

  group('sanitize, never reject — a bad route never costs a message', () {
    const Map<String, String?> cases = <String, String?>{
      'null': null,
      'an empty string': '',
      'whitespace only': '   ',
      'an unrooted path': 'jobs/apply',
      'a full URL': 'https://badabhai.ai/jobs',
      // A rooted path by every naive check, and a SCHEME-RELATIVE URL to a host
      // we do not control — rendered as a link on the admin screen it is an
      // outbound navigation an attacker chose.
      'a scheme-relative URL': '//evil.example/jobs',
      'a doubled slash anywhere': '/jobs//apply',
      'markup': '/jobs/<script>alert(1)</script>',
      'whitespace inside': '/jobs/my job',
      'percent-encoding': '/jobs/%2e%2e%2fadmin',
      'a newline': '/jobs\n/admin',
      'non-ASCII': '/नौकरी/आवेदन',
    };
    cases.forEach((String label, String? raw) {
      test('returns null for $label', () {
        expect(normalizeScreenContext(raw), isNull);
      });
    });

    test('null past the bound rather than a truncation', () {
      // A truncated route pattern names a screen nobody can navigate to and
      // looks authoritative on the admin list. Null reads as "unknown screen".
      expect(
        normalizeScreenContext('/${'a' * kWorkerFeedbackScreenMax}'),
        isNull,
      );
      expect(
        normalizeScreenContext('/${'a' * (kWorkerFeedbackScreenMax - 1)}'),
        isNotNull,
      );
    });

    test('the bound is measured AFTER substitution', () {
      // Four uuid segments is 148 raw characters — over the bound — and 20 once
      // normalized. The deep screens are exactly the ones a "button kaam nahi kar
      // raha" report is about, so measuring before substitution would discard the
      // reports that matter most.
      final String raw = '/a/6f2c04e0-4f89-41d3-9a0c-0305e82c3301' * 4;
      expect(raw.length, greaterThan(kWorkerFeedbackScreenMax));
      expect(normalizeScreenContext(raw), '/a/:id/a/:id/a/:id/a/:id');
    });

    test('never throws', () {
      for (final String? raw in <String?>[
        null,
        '/',
        '?',
        '#',
        '/' * 5000,
      ]) {
        expect(() => normalizeScreenContext(raw), returnsNormally);
      }
    });
  });

  group('what it PERMITS — the half a strip-everything bug would pass', () {
    test('the root, a plain route, a trailing slash, and surrounding spaces', () {
      expect(normalizeScreenContext('/'), '/');
      expect(normalizeScreenContext('/home'), '/home');
      expect(normalizeScreenContext('/jobs/'), '/jobs/');
      expect(normalizeScreenContext('  /settings/notifications  '),
          '/settings/notifications');
    });

    test('REAL worker-app routes reach the admin unchanged', () {
      // If these did not survive, every feedback row would read "unknown screen"
      // and the whole change would be inert while looking like it shipped.
      const List<String> routes = <String>[
        '/resume',
        '/jobs',
        '/profile',
        '/chat',
        '/alerts',
        '/settings',
        '/applied',
        '/invite',
        '/consent',
        '/name',
      ];
      for (final String route in routes) {
        expect(normalizeScreenContext(route), route, reason: route);
      }
    });

    test('a route pattern that already carries :id is left alone', () {
      // `:` is in the charset for exactly one reason — the substitution marker —
      // so a value that has already been normalized must be idempotent.
      expect(normalizeScreenContext('/jobs/:id/apply'), '/jobs/:id/apply');
      expect(
        normalizeScreenContext(normalizeScreenContext('/orders/91723')),
        '/orders/:id',
      );
    });

    test('a hyphenated, dotted or capitalised segment is structure, not an id',
        () {
      expect(normalizeScreenContext('/jobs/detail/job-1'), '/jobs/detail/job-1');
      expect(normalizeScreenContext('/pin/set'), '/pin/set');
      // The shared charset is `[A-Za-z0-9._:/-]` — dot, UNDERSCORE, colon,
      // slash, hyphen. Read it wrong in either language and the two halves
      // disagree about a perfectly ordinary route, which is precisely the drift
      // this file exists to catch.
      expect(normalizeScreenContext('/a.b/C_d-e'), '/a.b/C_d-e');
    });
  });
}
